Add handling to select CRTM cloud optical table based on cloud scheme and update calcanal_gfs.py

emilyhcliu opened the pull request:
# Description

This PR proposes updates for the following two scripts: 

1. In **scripts/exglobal_atmos_analysis.sh** --- Add handling to select CRTM cloud optical table based on cloud microphysical scheme indicated by `imp_physics'
The default scheme in the GFS forecast model is Thompson scheme (imp_physics = 8). 

2. In **/ush/calcanl_gfs.py** --- Increase the MPI number declared in the script due to increased variables to interplate increments and calculate analysis in the netcdf_io routines in GSI-utils.
Here is the related [PR #46  for GSI-utils](https://github.com/NOAA-EMC/GSI-utils/pull/46).  

# Type of change
- Maintenance (add handling)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
<!-- Please list any test you conducted, including the machine.
Cycled test on HERA

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

emilyhcliu committed:
Add handling to select CRTM cloud optical table based on cloud scheme.

aerorahul commented on scripts/exglobal_atmos_analysis.sh:
```diff
@@ -430,7 +430,19 @@ ${NLN} ${CRTM_FIX}/NPOESS.VISsnow.EmisCoeff.bin  ./crtm_coeffs/NPOESS.VISsnow.Em
 ${NLN} ${CRTM_FIX}/AerosolCoeff.bin              ./crtm_coeffs/AerosolCoeff.bin
-${NLN} ${CRTM_FIX}/CloudCoeff.GFDLFV3.-109z-1.bin ./crtm_coeffs/CloudCoeff.bin
+${NLN} ${CRTM_FIX}/CloudCoeff.bin                ./crtm_coeffs/CloudCoeff.bin
+if [[ "$imp_physics" = "8" ]]; then
+   echo "using CRTM Thompson cloud optical table"
+   ${NLN} ${CRTM_FIX}/CloudCoeff.Thompson08.-109z-1.bin ./crtm_coeffs/CloudCoeff.bin
+elif [[ "$imp_physics" = "11" ]]; then
+   echo "using CRTM GFDL cloud optical table"
+   ${NLN} ${CRTM_FIX}/CloudCoeff.GFDLFV3.-109z-1.bin ./crtm_coeffs/CloudCoeff.bin
+else
+   echo "INVALID imp_physics = $imp_physics"
+   echo "FATAL ERROR: No valid CRTM cloud optical table found for imp_physics =  " ${imp_physics}
+   exit 1
+fi
```
```suggestion
if (( imp_physics == 8 )); then
   echo "using CRTM Thompson cloud optical table"
   ${NLN} "${CRTM_FIX}/CloudCoeff.Thompson08.-109z-1.bin" ./crtm_coeffs/CloudCoeff.bin
elif (( imp_physics == 11 )); then
   echo "using CRTM GFDL cloud optical table"
   ${NLN} "${CRTM_FIX}/CloudCoeff.GFDLFV3.-109z-1.bin" ./crtm_coeffs/CloudCoeff.bin
else
   echo "INVALID imp_physics = ${imp_physics}"
   echo "FATAL ERROR: No valid CRTM cloud optical table found for imp_physics =  ${imp_physics}"
   exit 1
fi
```
Some simple edits will fix the shellcheck errors.
Otherwise looks good.

emilyhcliu committed:
Update scripts/exglobal_atmos_analysis.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

emilyhcliu commented:
@aerorahul Thanks for your suggested changes.  I have committed your suggestions.  Thanks a lot!

aerorahul commented on scripts/exglobal_atmos_analysis.sh:
```diff
@@ -430,7 +430,19 @@ ${NLN} ${CRTM_FIX}/NPOESS.VISsnow.EmisCoeff.bin  ./crtm_coeffs/NPOESS.VISsnow.Em
 ${NLN} ${CRTM_FIX}/FASTEM6.MWwater.EmisCoeff.bin ./crtm_coeffs/FASTEM6.MWwater.EmisCoeff.bin
 ${NLN} ${CRTM_FIX}/AerosolCoeff.bin              ./crtm_coeffs/AerosolCoeff.bin
-${NLN} ${CRTM_FIX}/CloudCoeff.GFDLFV3.-109z-1.bin ./crtm_coeffs/CloudCoeff.bin
+${NLN} ${CRTM_FIX}/CloudCoeff.bin                ./crtm_coeffs/CloudCoeff.bin
```
I think this line needs to be removed!

emilyhcliu replied:
OK.  Got it

emilyhcliu committed:
Increase the total number of MPI due to increase number of variables for
analysis increment interpolation and analysis calculation in
gsi_utils/src/netcdf_io

emilyhcliu committed:
Remove pre-defined cloud optical table

aerorahul reviewed (approved):
looks good.
I assume these files are available in the `CRTM_FIX` directory already and this PR does not need an update to the CRTM installations on the HPC's.

emilyhcliu commented:
> looks good. I assume these files are available in the `CRTM_FIX` directory already and this PR does not need an update to the CRTM installations on the HPC's.

Yes. They are already installed on the HPCs.

```
fixcrtm=/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/unified-env-rocky8/install/intel/2021.5.0/crtm-fix-2.4.0.1_emc-bm46d3q/fix

-rw-r--r-- 1 role.epic nems 1654180 Feb 14 17:55 CloudCoeff.bin
-rw-r--r-- 1 role.epic nems  359236 Feb 14 17:55 CloudCoeff.GFDLFV3.-109z-1.bin
-rw-r--r-- 1 role.epic nems  385887 Feb 14 17:55 CloudCoeff.GFDLFV3.-109z-1.nc4
-rw-r--r-- 1 role.epic nems  526609 Feb 14 17:55 CloudCoeff_MW_2.4.0_a.nc4
-rw-r--r-- 1 role.epic nems  875455 Feb 14 17:55 CloudCoeff_MW_2.4.0_exp_freq.nc4
-rw-r--r-- 1 role.epic nems  775536 Feb 14 17:55 CloudCoeff_MW_2.4.0_exp.nc4
-rw-r--r-- 1 role.epic nems 1682681 Feb 14 17:55 CloudCoeff.nc4
-rw-r--r-- 1 role.epic nems  897462 Feb 14 17:55 CloudCoeff_TAMU-07Sep2015.nc4
-rw-r--r-- 1 role.epic nems 1654180 Feb 14 17:55 CloudCoeff_TAMU-11Sep2014.bin
-rw-r--r-- 1 role.epic nems  686282 Feb 14 17:55 CloudCoeff_TAMU-11Sep2014.nc4
-rw-r--r-- 1 role.epic nems  359236 Feb 14 17:55 CloudCoeff.Thompson08.-109z-1.bin
-rw-r--r-- 1 role.epic nems  385887 Feb 14 17:55 CloudCoeff.Thompson08.-109z-1.nc4
-rw-r--r-- 1 role.epic nems  359236 Feb 14 17:55 CloudCoeff.WSM6.-109z-1.bin
-rw-r--r-- 1 role.epic nems  385887 Feb 14 17:55 CloudCoeff.WSM6.-109z-1.nc4
-rw-r--r-- 1 role.epic nems  162932 Feb 14 17:55 CloudCoeff.WSM6.MieSphere.bin
-rw-r--r-- 1 role.epic nems  189583 Feb 14 17:55 CloudCoeff.WSM6.MieSphere.nc4

```

emilyhcliu committed:
Merge branch 'feature/calc_analysis' into feature/crtm_cloud_coefficient

emilyhcliu commented:
@aerorahul. I combined [PR 2648](https://github.com/NOAA-EMC/global-workflow/pull/2648) into this PR.  Changes are merged as well.

WalterKolczynski-NOAA commented on ush/calcanl_gfs.py:
```diff
@@ -248,7 +248,7 @@ def calcanl_gfs(DoIAU, l4DEnsVar, Write4Danl, ComOut, APrefix,
                         for a in range(0, 5):
                             hostfile.write(hosts[ihost] + '\n')
-                    for a in range(0, 9):  # need 9 more of the same host for the 10 tasks for chgres_inc
+                    for a in range(0, 12):  # need 9 more of the same host for the 10 tasks for chgres_inc
```
```suggestion
                    for a in range(0, 12):  # need 12 more of the same host for the 13 tasks for chgres_inc
```

WalterKolczynski-NOAA commented on ush/calcanl_gfs.py:
```diff
@@ -135,7 +135,7 @@ def calcanl_gfs(DoIAU, l4DEnsVar, Write4Danl, ComOut, APrefix,
     ExecCMDMPI = ExecCMDMPI.replace("$ncmd", str(nFH))
     ExecCMDLevs = ExecCMDMPI.replace("$ncmd", str(levs))
-    ExecCMDMPI10 = ExecCMDMPI.replace("$ncmd", str(10))
+    ExecCMDMPI13 = ExecCMDMPI.replace("$ncmd", str(13))
```
Outside the scope of this PR, but it looks like this script is in need of a refactor to remove these hard-coded values that need to be changed in a half-dozen places (and to make it more pythonic in general).

emilyhcliu replied:
@WalterKolczynski-NOAA  Agreed!

WalterKolczynski-NOAA reviewed (changes requested):
Just updating a comment to match the change made. A refactor is needed here, but outside scope for this PR.

WalterKolczynski-NOAA commented:
Missed updating a variable name:
```
Interpolating increment for f003
interp_inc 3 OrderedDict([('setup', {'lon_out': 384, 'lat_out': 192, 'lev': 127, 'infile': "'siginc.nc.03'", 'outfile': "'inc.fullres.03'"})])
Traceback (most recent call last):
  File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2645/global-workflow/ush/calcanl_gfs.py", line 366, in <module>
    calcanl_gfs(DoIAU, l4DEnsVar, Write4Danl, ComOut, APrefix,
  File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2645/global-workflow/ush/calcanl_gfs.py", line 256, in calcanl_gfs
    job = subprocess.Popen(ExecCMDMPI10_host + ' ' + CalcAnlDir + '/chgres_inc.x', shell=True, cwd=CalcAnlDir)
NameError: name 'ExecCMDMPI10_host' is not defined
+ exglobal_atmos_analysis_calc.sh[1]: postamble exglobal_atmos_analysis_calc.sh 1717275805 1
```

aerorahul commented:
@emilyhcliu 
Does this PR require https://github.com/NOAA-EMC/GSI-utils/pull/46 to be merged?

emilyhcliu committed:
Update ush/calcanl_gfs.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

emilyhcliu commented:
> @emilyhcliu Does this PR require [NOAA-EMC/GSI-utils#46](https://github.com/NOAA-EMC/GSI-utils/pull/46) to be merged?

@aerorahul The merge of GSI-utils PR #46  requires the MPI number change from this PR.  
But, I am not sure if this PR requires GSI-utils PR #46
The MPI numbers (10 vs. 13) specified in the `calcanl_gfs.py` seem to be hardwired with the code in netcdf_io in GSI-utils

I am checking the failed log and see if I can found any clues.

emilyhcliu commented:
I found a typo.  I will commit the fix

emilyhcliu committed:
Update calcanl_gfs.py

fix typo: from ExecCMDMPI10_host to ExecCMDMPI13_host

emilyhcliu committed:
Merge branch 'develop' into feature/crtm_cloud_coefficient

emilyhcliu commented:
> Missed updating a variable name:
> 
> ```
> Interpolating increment for f003
> interp_inc 3 OrderedDict([('setup', {'lon_out': 384, 'lat_out': 192, 'lev': 127, 'infile': "'siginc.nc.03'", 'outfile': "'inc.fullres.03'"})])
> Traceback (most recent call last):
>   File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2645/global-workflow/ush/calcanl_gfs.py", line 366, in <module>
>     calcanl_gfs(DoIAU, l4DEnsVar, Write4Danl, ComOut, APrefix,
>   File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2645/global-workflow/ush/calcanl_gfs.py", line 256, in calcanl_gfs
>     job = subprocess.Popen(ExecCMDMPI10_host + ' ' + CalcAnlDir + '/chgres_inc.x', shell=True, cwd=CalcAnlDir)
> NameError: name 'ExecCMDMPI10_host' is not defined
> + exglobal_atmos_analysis_calc.sh[1]: postamble exglobal_atmos_analysis_calc.sh 1717275805 1
> ```

@WalterKolczynski-NOAA Thanks for finding the missed update.  I just committed the fix.

emilyhcliu commented:
Great!  All checks passed..

emilyhcliu commented:
> Great! All checks passed..

@WalterKolczynski-NOAA Could you verify if all CI tests passed and update the labels accordingly?  Thanks.

emilyhcliu commented:
> > Great! All checks passed..
> 
> @WalterKolczynski-NOAA Could you verify if all CI tests passed and update the labels accordingly? Thanks.

@WalterKolczynski-NOAA I saw a green check indicated that all tests passed.  Can I remove the failed lables added to this PR?

emilyhcliu commented:
@WalterKolczynski-NOAA Thanks!!

WalterKolczynski-NOAA commented:
Going to merge without passing on MSU machines.
